File: backend/src/services/companyLookupService.ts
import { HttpError } from "../utils/http.js";

function normalizeCnpj(value: unknown) {
  const cnpj = String(value ?? "").replace(/\D/g, "");
  if (!/^\d{14}$/.test(cnpj)) throw new HttpError(422, "Informe um CNPJ válido.");
  return cnpj;
}

function text(value: unknown, maxLength: number) {
  return typeof value === "string" ? value.trim().slice(0, maxLength) : "";
}

export async function lookupCompanyAddress(value: unknown) {
  const cnpj = normalizeCnpj(value);
  let response: Response;
  try {
    response = await fetch(`https://brasilapi.com.br/api/cnpj/v1/${cnpj}`, {
      signal: AbortSignal.timeout(5_000),
    });
  } catch {
    throw new HttpError(503, "Não foi possível confirmar o endereço pelo CNPJ agora.");
  }

  if (response.status === 404) throw new HttpError(404, "CNPJ não encontrado.");
  if (!response.ok) throw new HttpError(503, "Não foi possível confirmar o endereço pelo CNPJ agora.");
  const data: unknown = await response.json().catch(() => null);
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    throw new HttpError(503, "Não foi possível confirmar o endereço pelo CNPJ agora.");
  }

  const company = data as Record<string, unknown>;
  const stateCode = text(company.uf, 2).toUpperCase();
  const city = text(company.municipio, 100);
  if (!city || !/^[A-Z]{2}$/.test(stateCode)) {
    throw new HttpError(404, "O CNPJ não possui endereço suficiente para confirmação.");
  }

  return {
    cnpj,
    postalCode: text(company.cep, 12).replace(/\D/g, "").slice(0, 8),
    street: text(company.logradouro, 160),
    number: text(company.numero, 40),
    complement: text(company.complemento, 120),
    neighborhood: text(company.bairro, 100),
    city,
    stateCode,
  };
}

File: site/backend/src/main/java/br/com/rodogarcia/site/backend/service/CompanyLookupService.java
package br.com.rodogarcia.site.backend.service;

import java.util.Locale;

import br.com.rodogarcia.site.backend.dto.response.CompanyAddressResponse;
import br.com.rodogarcia.site.backend.exception.ApiException;
import br.com.rodogarcia.site.backend.integration.ProviderAccessException;
import br.com.rodogarcia.site.backend.integration.ProviderHttpResponse;
import br.com.rodogarcia.site.backend.integration.brasilapi.BrasilApiClient;
import br.com.rodogarcia.site.backend.validation.StrictJson;
import org.springframework.stereotype.Service;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

@Service
public class CompanyLookupService {

    private static final String UNAVAILABLE_MESSAGE =
        "Não foi possível confirmar o endereço pelo CNPJ agora.";

    private final BrasilApiClient client;
    private final JsonMapper jsonMapper;

    public CompanyLookupService(BrasilApiClient client, JsonMapper jsonMapper) {
        this.client = client;
        this.jsonMapper = jsonMapper;
    }

    public CompanyAddressResponse lookup(String value) {
        String cnpj = normalizeCnpj(value);
        ProviderHttpResponse response;
        try {
            response = client.lookupCompany(cnpj);
        } catch (ProviderAccessException error) {
            throw new ApiException(503, UNAVAILABLE_MESSAGE);
        }

        if (response.status() == 404) {
            throw new ApiException(404, "CNPJ não encontrado.");
        }
        if (response.status() < 200 || response.status() >= 300) {
            throw new ApiException(503, UNAVAILABLE_MESSAGE);
        }

        JsonNode company = parseJson(response.body());
        if (company == null || !company.isObject()) {
            throw new ApiException(503, UNAVAILABLE_MESSAGE);
        }

        String stateCode = text(company.get("uf"), 2).toUpperCase(Locale.ROOT);
        String city = text(company.get("municipio"), 100);
        if (city.isEmpty() || !stateCode.matches("^[A-Z]{2}$")) {
            throw new ApiException(404, "O CNPJ não possui endereço suficiente para confirmação.");
        }

        return new CompanyAddressResponse(
            cnpj,
            digits(text(company.get("cep"), 12), 8),
            text(company.get("logradouro"), 160),
            text(company.get("numero"), 40),
            text(company.get("complemento"), 120),
            text(company.get("bairro"), 100),
            city,
            stateCode
        );
    }

    private JsonNode parseJson(byte[] body) {
        try {
            return StrictJson.readTree(jsonMapper, body);
        } catch (JacksonException error) {
            return null;
        }
    }

    private static String normalizeCnpj(String value) {
        String cnpj = (value == null ? "" : value).replaceAll("\\D", "");
        if (!cnpj.matches("^\\d{14}$")) {
            throw new ApiException(422, "Informe um CNPJ válido.");
        }
        return cnpj;
    }

    private static String text(JsonNode value, int maximumLength) {
        if (value == null || !value.isString()) {
            return "";
        }
        String normalized = NodeStringCompatibility.trim(value.stringValue());
        return normalized.length() <= maximumLength
            ? normalized
            : normalized.substring(0, maximumLength);
    }

    private static String digits(String value, int maximumLength) {
        String normalized = value.replaceAll("\\D", "");
        return normalized.length() <= maximumLength
            ? normalized
            : normalized.substring(0, maximumLength);
    }
}
